src/how-to-build-and-run.md

```diff
@@ -43,54 +43,237 @@ debuginfo = true
+| copy (uplift) `stage0-sysroot` to `stage1`          | `build/HOST/stage1/lib`               |
+| `stage1` builds `libstd`                            | `build/HOST/stage1-std/TARGET`        |
+| copy `stage1-std` (HOST only)                       | `build/HOST/stage1/lib/rustlib/HOST`  |
+| `stage1` builds `libtest`                           | `build/HOST/stage1-test/TARGET`       |
+| copy `stage1-test` (HOST only)                      | `build/HOST/stage1/lib/rustlib/HOST`  |
+| `stage1` builds `rustc`                             | `build/HOST/stage1-rustc/HOST`        |
+| copy `stage1-rustc` (except executable)             | `build/HOST/stage1/lib/rustlib/HOST`  |
+| `stage1` builds `codegen`                           | `build/HOST/stage1-codegen/HOST`      |
+
+`--stage=1` stops here.
+
+| Stage 2 Action                            | Output                                                          |
+|-------------------------------------------|-----------------------------------------------------------------|
+| copy (uplift) `stage1-rustc` executable   | `build/HOST/stage2/bin`                                         |
+| copy (uplift) `stage1-sysroot`            | `build/HOST/stage2/lib and build/HOST/stage2/lib/rustlib/HOST`  |
+| `stage2` builds `libstd` (except HOST?)   | `build/HOST/stage2-std/TARGET`                                  |
+| copy `stage2-std` (not HOST targets)      | `build/HOST/stage2/lib/rustlib/TARGET`                          |
+| `stage2` builds `libtest` (except HOST?)  | `build/HOST/stage2-test/TARGET`                                 |
+| copy `stage2-test` (not HOST targets)     | `build/HOST/stage2/lib/rustlib/TARGET`                          |
+| `stage2` builds `rustdoc`                 | `build/HOST/stage2-tools/HOST`                                  |
+| copy `rustdoc`                            | `build/HOST/stage2/bin`                                         |
+
+`--stage=2` stops here.
+
+Note that the convention `x.py` uses is that:
+- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts".
+- A `--stage N` flag means build _with_ stage N.
+
+In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which
+will later be uplifted to stage1_.
+
+Every time any of the main artifacts (`std`, `test`, `rustc`) are compiled, two
+steps are performed.
+When `std` is compiled by a stage N compiler, that `std` will be linked to
+programs built by the stage N compiler (including test and `rustc` built later
+on). It will also be used by the stage (N+1) compiler to link against itself.
+This is somewhat intuitive if one thinks of the stage (N+1) compiler as "just"
+another program we are building with the stage N compiler. In some ways, `rustc`
+(the binary, not the `rustbuild` step) could be thought of as one of the few
```

Comment:
I'm not sure I understand this sentence.  `rustc` links to libstd/libcore?

Reply:
Yes, but IIUC, bootstrap treats libstd/libcore as "just another crate" that we are linking to our program. In other words, one can think of the compiler as a `no_core` program that cross-compiles and links with its own versions of the standard library.